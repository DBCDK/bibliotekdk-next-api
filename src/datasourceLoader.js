import DataLoader from "dataloader";
import { log } from "dbc-node-logger";
import { withRedis } from "./datasources/redis.datasource";
import monitor from "./utils/monitor";
import { getFilesRecursive } from "./utils/utils";

// Find all datasources in src/datasources
export const datasources = getFilesRecursive("./src/datasources")
  .map((file) => {
    if (!file.path.endsWith(".datasource.js")) {
      return;
    }
    const { load, options, batchLoader, status } = require(file.path);
    if (!load) {
      return;
    }

    // Extract datasource name from filename
    const name = file.file.replace(".datasource.js", "");

    // Monitor the load function from the datasource
    const monitoredLoad = monitor(
      {
        name: `REQUEST_${name}`,
        help: `${name} requests`,
      },
      load
    );

    // if datasource exports a createBatchLoader we use that,
    // otherwise we use default batch loader
    let monitoredBatchLoader = batchLoader
      ? (keys) => batchLoader(keys, monitoredLoad)
      : async (keys) => {
          return await Promise.all(keys.map((key) => monitoredLoad(key)));
        };

    // Check if Redis is configured for this datasource
    if (options && options.redis && options.redis.prefix && options.redis.ttl) {
      monitoredBatchLoader = withRedis(monitoredBatchLoader, {
        prefix: options.redis.prefix,
        ttl: options.redis.ttl,
      });
    }

    const statusChecker = status && (async () => await status(monitoredLoad));

    return {
      batchLoader: monitoredBatchLoader,
      name,
      options,
      statusChecker,
    };
  })
  .filter((func) => !!func);

log.info(`found ${datasources.length} datasources`, {
  datasources: datasources.map((datasource) => ({
    name: datasource.name,
    options: datasource.options,
  })),
});

/**
 * Will instantiate dataloaders from datasources
 */
export default function createDataLoaders() {
  const result = {};
  datasources.forEach((datasource) => {
    result[datasource.name] = new DataLoader(datasource.batchLoader, {
      // If key is an object, we stringify
      // to make it useful as a cache key
      cacheKeyFn: (key) =>
        typeof key === "object" ? JSON.stringify(key) : key,
    });
  });
  return result;
}
