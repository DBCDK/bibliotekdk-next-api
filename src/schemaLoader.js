/**
 * @file Responsible for loading the schema
 *
 * It will check all files recursively in ./src/schema
 * and load type definitions and resolvers
 */

import { makeExecutableSchema, mergeSchemas } from "graphql-tools";
import drupalSchema from "./schema/external/drupal";
import { log } from "dbc-node-logger";
import { getFilesRecursive } from "./utils/utils";

/**
 * Will check all files in schema folder
 * for type definitions and resolvers.
 *
 * These are then loaded
 */
function schemaLoader() {
  let allTypeDefs = [];
  let allResolvers = {};

  // Load files in schema folder
  const files = getFilesRecursive("./src/schema");

  // Require typeDefs and resolvers
  files.forEach((file) => {
    if (!file.path.endsWith(".js")) {
      return;
    }
    const { typeDef, resolvers } = require(file.path);
    if (typeDef) {
      allTypeDefs = [...allTypeDefs, typeDef];
      log.info(`Found type definition in ${file.path}`);
    }
    if (resolvers) {
      allResolvers = { ...allResolvers, ...resolvers };
    }
  });

  return { typeDefs: allTypeDefs, resolvers: allResolvers };
}

const { typeDefs, resolvers } = schemaLoader();

/**
 * Create executable schema from type definitions and resolvers
 */
export const internalSchema = makeExecutableSchema({
  typeDefs,
  resolvers,
});

/**
 * We stitch the internal schema together with the external drupal schema
 */
export default async () => {
  return mergeSchemas({
    subschemas: [{ schema: await drupalSchema() }, { schema: internalSchema }],
  });
};
