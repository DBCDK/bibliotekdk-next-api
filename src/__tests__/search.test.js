/**
 * @file This file contains tests for the search operation
 */

import { graphql } from "graphql";
import { internalSchema } from "../schemaLoader";
import { createMockedDataLoaders } from "../datasourceLoader";

export async function performTestQuery({ query, variables, context }) {
  return graphql(internalSchema, query, null, context, variables);
}

test("search - with all filters and facets", async () => {
  const result = await performTestQuery({
    query: `
      {
        search(q: {all: "*", title: "*", creator: "*", subject: "*"},filters: {accessType: [], audience: [], creator: [], fictionNonfiction: [], fictiveCharacter: [], genre: [], language: [], materialType: [], subject: [], workType: []}) {
          hitcount
          works(offset: 0, limit: 10) {
            id
          }
          facets(facets: [workType, language, materialType, fictiveCharacter, genre, audience, accessType, fictionNonfiction, subject, creator]) {
            name
            values(limit: 1) {
              term
              count
            }
          }
        }
      }
        `,
    variables: {},
    context: { datasources: createMockedDataLoaders() },
  });
  expect(result).toMatchSnapshot();
});
