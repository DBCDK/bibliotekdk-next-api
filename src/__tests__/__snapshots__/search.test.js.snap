// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`search - with all filters and facets 1`] = `
Object {
  "data": Object {
    "search": Object {
      "facets": Array [
        Object {
          "name": "workType",
          "values": Array [
            Object {
              "count": 2119328,
              "term": "Artikler",
            },
          ],
        },
        Object {
          "name": "language",
          "values": Array [
            Object {
              "count": 2252441,
              "term": "Dansk",
            },
          ],
        },
        Object {
          "name": "materialType",
          "values": Array [
            Object {
              "count": 1148775,
              "term": "tidsskriftsartikel",
            },
          ],
        },
        Object {
          "name": "fictiveCharacter",
          "values": Array [
            Object {
              "count": 413,
              "term": "Batman",
            },
          ],
        },
        Object {
          "name": "genre",
          "values": Array [
            Object {
              "count": 47372,
              "term": "rock",
            },
          ],
        },
        Object {
          "name": "audience",
          "values": Array [
            Object {
              "count": 3253857,
              "term": "voksenmaterialer",
            },
          ],
        },
        Object {
          "name": "accessType",
          "values": Array [
            Object {
              "count": 3255553,
              "term": "Fysisk",
            },
          ],
        },
        Object {
          "name": "fictionNonfiction",
          "values": Array [
            Object {
              "count": 2399695,
              "term": "Faglitteratur",
            },
          ],
        },
        Object {
          "name": "subject",
          "values": Array [
            Object {
              "count": 105800,
              "term": "historie",
            },
          ],
        },
        Object {
          "name": "creator",
          "values": Array [
            Object {
              "count": 6209,
              "term": "Association for Computing Machinery-Digital Library",
            },
          ],
        },
      ],
      "hitcount": 1,
      "works": Array [
        Object {
          "id": "work-of:870970-basis:53557791",
        },
      ],
    },
  },
}
`;
